pkg/test/proxy/envoy/envoy.go

```diff
@@ -0,0 +1,212 @@
+		return err
+	}
+
+	envoyPath := e.BinPath
+	if envoyPath == "" {
+		// No binary specified, assume a default location under ISTIO_OUT
+		var err error
+		envoyPath, err = getDefaultEnvoyBinaryPath()
+		if err != nil {
+			return err
+		}
+	}
+
+	// Run the envoy binary
+	args := e.getCommandArgs()
+	e.cmd = exec.Command(envoyPath, args...)
+	e.cmd.Stderr = os.Stderr
+	e.cmd.Stdout = os.Stdout
+	if err := e.cmd.Start(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Stop kills the Envoy process.
+func (e *Envoy) Stop() error {
+	if e.cmd == nil || e.cmd.Process != nil {
+		// Wasn't previously started - nothing to do.
+		return nil
+	}
+
+	// Kill the process.
+	return e.cmd.Process.Kill()
+}
+
+func (e *Envoy) validateCommandArgs() error {
+	if e.BinPath != "" {
+		// Ensure the binary exists.
+		if err := checkFileExists(e.BinPath); err != nil {
+			return fmt.Errorf("specified Envoy binary does not exist: %s", e.BinPath)
+		}
+	}
+	if e.ConfigFile == "" {
+		return fmt.Errorf("configFile must be specified before running Envoy")
+	}
+	return nil
+}
+
+func (e *Envoy) getCommandArgs() []string {
+	// We need to make sure each envoy has a unique base ID in order to run multiple instances on the same
+	// machine.
+	baseID := rand.Intn(maxBaseID) + 1
+
+	// Prefix Envoy log entries with [ENVOY] to make them distinct from other logs if mixed within the same stream (e.g. stderr)
+	logFormat := "[ENVOY] [%Y-%m-%d %T.%e][%t][%l][%n] %v"
+
+	args := []string{
+		"--base-id",
+		strconv.Itoa(baseID),
+		// Always force v2 config.
+		"--v2-config-only",
+		"--config-path",
+		e.ConfigFile,
+		"--log-level",
+		string(e.getLogLevel()),
+		"--log-format",
+		logFormat,
+	}
+
+	if e.LogFilePath != "" {
+		args = append(args, "--log-path", e.LogFilePath)
+	}
+	return args
+}
+
+func (e *Envoy) getLogLevel() LogLevel {
+	if e.LogLevel != "" {
+		return e.LogLevel
+	}
+	return DefaultLogLevel
+}
+
+func checkFileExists(f string) error {
+	if _, err := os.Stat(f); os.IsNotExist(err) {
+		return err
+	}
+	return nil
+}
+
+func getDefaultEnvoyBinaryPath() (string, error) {
+	istioOut, err := getIstioOut()
+	if err != nil {
+		return "", err
+	}
+
+	// Find all of the debug envoy binaries.
+	binDir := filepath.Join(istioOut, "debug")
+	binPrefix := filepath.Join(binDir, "envoy-debug-")
+	binPaths, err := filepath.Glob(binPrefix + "*")
+	if err != nil {
+		return "", err
+	}
+
+	if len(binPaths) == 0 {
+		return "", fmt.Errorf("unable to locate Envoy binary in dir %s", istioOut)
+	}
+
+	// Find the most recent debug binary.
+	latestBinPath := ""
+	latestBinTime := int64(0)
+	for _, binPath := range binPaths {
+		binFile, err := os.Stat(binPath)
+		if err != nil {
+			// Should never happen
+			return "", err
+		}
+		binTime := binFile.ModTime().Unix()
+		if binTime > latestBinTime {
+			latestBinTime = binTime
+			latestBinPath = binPath
+		}
+	}
+
+	return latestBinPath, nil
+}
+
+func getIstioOut() (string, error) {
```

Comment:
We already have utilities for this in test/utils

Reply:
OK, it's in a strange location however ... it's in `tests/util/pilot_server.go`.  I'll create a version of this in `pkg/test/util.go` so that it's in a more common place. Keep in mind that I'm merging into the `test-framework` branch, so this won't be making into master anytime soon.